File: src/aais_tasks/aais_task_store.py
"""Durable AAIS task store under .runtime/aais_tasks/tasks.json.

# Mythic: AAIS task ledger
# Engineering: AaisTaskStore
"""

from __future__ import annotations

import json
import os
import threading
from pathlib import Path
from typing import Any

from src.aais_tasks.aais_task_model import AaisTask, TaskStatus, new_task, _utc_now


def _default_path(runtime_root: Path | None = None) -> Path:
    if runtime_root is not None:
        return Path(runtime_root) / "aais_tasks" / "tasks.json"
    configured = os.getenv("AAIS_RUNTIME_DIR")
    if configured:
        return Path(configured).expanduser() / "aais_tasks" / "tasks.json"
    return Path(__file__).resolve().parents[2] / ".runtime" / "aais_tasks" / "tasks.json"


class AaisTaskStore:
    def __init__(self, file_path: Path | None = None, *, runtime_root: Path | None = None) -> None:
        self._path = file_path or _default_path(runtime_root)
        self._lock = threading.Lock()

    def _load(self) -> list[AaisTask]:
        if not self._path.is_file():
            return []
        try:
            raw = json.loads(self._path.read_text(encoding="utf-8"))
            rows = raw.get("tasks") if isinstance(raw, dict) else []
            if not isinstance(rows, list):
                return []
            return [AaisTask.from_dict(r) for r in rows if isinstance(r, dict)]
        except (json.JSONDecodeError, OSError):
            return []

    def _save(self, tasks: list[AaisTask]) -> None:
        self._path.parent.mkdir(parents=True, exist_ok=True)
        payload = {
            "updatedAt": _utc_now(),
            "tasks": [t.to_dict() for t in tasks],
        }
        self._path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")

    def create(self, **kwargs: Any) -> AaisTask:
        with self._lock:
            tasks = self._load()
            task = new_task(
                str(kwargs.get("title") or "Untitled"),
                description=kwargs.get("description"),
                due_date=kwargs.get("due_date") or kwargs.get("dueDate"),
                status=kwargs.get("status") or "notStarted",
                priority=kwargs.get("priority") or "normal",
                tags=list(kwargs.get("tags") or []),
                source=kwargs.get("source") or "aais",
                graph_id=kwargs.get("graph_id") or kwargs.get("graphId"),
            )
            tasks.append(task)
            self._save(tasks)
            return task

    def list(self) -> list[AaisTask]:
        with self._lock:
            return self._load()

    def get(self, task_id: str) -> AaisTask | None:
        with self._lock:
            for t in self._load():
                if t.id == task_id:
                    return t
            return None

    def update(self, task_id: str, **patch: Any) -> AaisTask | None:
        with self._lock:
            tasks = self._load()
            for i, t in enumerate(tasks):
                if t.id != task_id:
                    continue
                data = t.to_dict()
                if "title" in patch and patch["title"] is not None:
                    data["title"] = str(patch["title"])
                if "description" in patch:
                    data["description"] = patch["description"]
                if "dueDate" in patch or "due_date" in patch:
                    data["dueDate"] = patch.get("dueDate", patch.get("due_date"))
                if "status" in patch and patch["status"]:
                    data["status"] = patch["status"]
                if "priority" in patch and patch["priority"]:
                    data["priority"] = patch["priority"]
                if "tags" in patch and patch["tags"] is not None:
                    data["tags"] = list(patch["tags"])
                if "source" in patch and patch["source"]:
                    data["source"] = patch["source"]
                if "graphId" in patch or "graph_id" in patch:
                    data["graphId"] = patch.get("graphId", patch.get("graph_id"))
                data["updatedAt"] = _utc_now()
                updated = AaisTask.from_dict(data)
                tasks[i] = updated
                self._save(tasks)
                return updated
            return None

    def update_status(self, task_id: str, status: TaskStatus) -> AaisTask | None:
        return self.update(task_id, status=status)

File: aais-middleware/src/aais_tasks/aais_task_store.ts
/**
 * Mythic: AAIS task ledger
 * Engineering: AaisTaskStore — durable JSON under .runtime/aais_tasks/
 */
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { randomUUID } from "node:crypto";
import {
  normalizeTask,
  type AaisTask,
  type CreateAaisTaskInput,
  type UpdateAaisTaskInput,
} from "./aais_task_model.js";

export interface AaisTaskStoreOptions {
  /** Absolute or cwd-relative path to tasks.json */
  filePath?: string;
  runtimeRoot?: string;
}

function defaultFilePath(runtimeRoot?: string): string {
  const root = runtimeRoot || process.env.AAIS_RUNTIME_DIR || join(process.cwd(), ".runtime");
  return join(root, "aais_tasks", "tasks.json");
}

export class AaisTaskStore {
  private readonly filePath: string;

  constructor(opts: AaisTaskStoreOptions = {}) {
    this.filePath = opts.filePath || defaultFilePath(opts.runtimeRoot);
  }

  private ensureDir(): void {
    mkdirSync(dirname(this.filePath), { recursive: true });
  }

  private load(): AaisTask[] {
    if (!existsSync(this.filePath)) return [];
    try {
      const raw = JSON.parse(readFileSync(this.filePath, "utf8")) as { tasks?: unknown };
      const list = Array.isArray(raw.tasks) ? raw.tasks : [];
      return list
        .filter((t): t is Record<string, unknown> => t != null && typeof t === "object")
        .map((t) =>
          normalizeTask({
            id: String(t.id || randomUUID()),
            title: String(t.title || "Untitled"),
            description: t.description != null ? String(t.description) : undefined,
            createdAt: t.createdAt != null ? String(t.createdAt) : undefined,
            dueDate: t.dueDate != null ? String(t.dueDate) : undefined,
            status: t.status as AaisTask["status"],
            priority: t.priority as AaisTask["priority"],
            tags: Array.isArray(t.tags) ? t.tags.map(String) : undefined,
            source: t.source as AaisTask["source"],
            graphId: t.graphId != null ? String(t.graphId) : undefined,
            updatedAt: t.updatedAt != null ? String(t.updatedAt) : undefined,
          }),
        );
    } catch {
      return [];
    }
  }

  private save(tasks: AaisTask[]): void {
    this.ensureDir();
    writeFileSync(
      this.filePath,
      JSON.stringify({ updatedAt: new Date().toISOString(), tasks }, null, 2) + "\n",
      "utf8",
    );
  }

  create(input: CreateAaisTaskInput): AaisTask {
    const tasks = this.load();
    const task = normalizeTask({
      id: randomUUID(),
      title: input.title,
      description: input.description,
      dueDate: input.dueDate,
      status: input.status || "notStarted",
      priority: input.priority,
      tags: input.tags,
      source: input.source || "aais",
      graphId: input.graphId,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    });
    tasks.push(task);
    this.save(tasks);
    return task;
  }

  list(): AaisTask[] {
    return this.load();
  }

  get(id: string): AaisTask | undefined {
    return this.load().find((t) => t.id === id);
  }

  update(id: string, patch: UpdateAaisTaskInput): AaisTask | undefined {
    const tasks = this.load();
    const idx = tasks.findIndex((t) => t.id === id);
    if (idx < 0) return undefined;
    const cur = tasks[idx]!;
    const next = normalizeTask({
      ...cur,
      title: patch.title ?? cur.title,
      description:
        patch.description !== undefined ? patch.description || undefined : cur.description,
      dueDate:
        patch.dueDate === null
          ? undefined
          : patch.dueDate !== undefined
            ? patch.dueDate
            : cur.dueDate,
      status: patch.status ?? cur.status,
      priority: patch.priority ?? cur.priority,
      tags: patch.tags ?? cur.tags,
      source: patch.source ?? cur.source,
      graphId:
        patch.graphId === null
          ? undefined
          : patch.graphId !== undefined
            ? patch.graphId
            : cur.graphId,
      createdAt: cur.createdAt,
      updatedAt: new Date().toISOString(),
    });
    tasks[idx] = next;
    this.save(tasks);
    return next;
  }
}
